src/basic/parse-util.c

```diff
@@ -480,13 +507,24 @@ int safe_atou16_full(const char *s, unsigned base, uint16_t *ret) {
         unsigned long l;
 
         assert(s);
-        assert(ret);
-        assert(base <= 16);
+        assert(SAFE_ATO_MASK_FLAGS(base) <= 16);
+
+        if (FLAGS_SET(base, SAFE_ATO_REFUSE_LEADING_WHITESPACE) &&
+            strchr(WHITESPACE, s[0]))
+                return -EINVAL;
 
         s += strspn(s, WHITESPACE);
 
+        if (FLAGS_SET(base, SAFE_ATO_REFUSE_PLUS_MINUS) &&
+            IN_SET(s[0], '+', '-'))
+                return -EINVAL;
+
+        if (FLAGS_SET(base, SAFE_ATO_REFUSE_LEADING_ZERO) &&
+            s[0] == '0' && s[1] != 0)
+                return -EINVAL;
+
         errno = 0;
-        l = strtoul(s, &x, base);
+        l = strtoul(s, &x, SAFE_ATO_MASK_FLAGS(base));
```

Comment:
Hmm, why do we need to check those conditions above, if we pass those flags to `strtoul`? I guess there's some reason for it, so maybe add a comment.

Reply:
we don't pass them to strtoul(). The`SAFE_ATO_MASK_FLAGS()` macro *masks* the flags and leaves only the base in the value

Reply:
Is this approach clean enough?

Reply:
Oh, OK. I missed the inversion in the macro.